### PR TITLE
Setup automated releases from GitHub Actions CI

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, releases were done from a local computer with the
sbt-release plugin. Now, releases happen automatically from GitHub
Actions. Every merge into master should trigger a SNAPSHOT release
and every git tag push should trigger a stable Sonatype release.

For more details, see https://github.com/olafurpg/sbt-ci-release

### CODE REVIEW

**Review by regadas — APPROVED** [review]
This is super nice! 🎉 I guess this just speeds up the process of moving completely to GHA as we did in chill.

Now @posco just needs to add a few secrets to the repo! 🚀 (I don't have enough perms)

**Comment by olafurpg** [comment]
I just received admin access for this repo and configured the Sonatype+PGP secrets. It may take several attempts after this is merged to validate that everything is working as expected.

> I guess this just speeds up the process of moving completely to GHA as we did in chill.

Yes, I think it would be a nice followup PR to move all the CI to GitHub Actions.

**Comment by olafurpg** [comment]
I'm planning to followup with similar changes in the chill and bijection repos.

File: version.sbt
Comment (olafurpg) [inline version.sbt:1]: This changes the local value of `version` to be formatted like this 0.13.7+34-2d333018-SNAPSHOT. Is that an acceptable change? I think it's possible to get configure the version to not include the git describe information you prefer to keep the old behavior.

Comment (regadas) [reply version.sbt:1]: imo i think it's ok.

**Comment by regadas** [comment]
Yeah following up with a PR would be nice... I've been meaning to do this for a long time lol

File: .github/workflows/release.yml
Comment (regadas) [inline .github/workflows/release.yml:11]: One i thing i realised is that we can add some artifact cache here

```
      - name: cache SBT
        uses: coursier/cache-action@v3
```

Comment (olafurpg) [reply .github/workflows/release.yml:11]: This job only runs the releases, which I think is desirable to run with a clean machine. My experience is that GitHub Actions is usually fast enough for most projects without caching

Comment (regadas) [reply .github/workflows/release.yml:11]: right, anecdotally I've seen it fail a few times ¯\_(ツ)_/¯ I guess we can always re-trigger?

Comment (olafurpg) [reply .github/workflows/release.yml:11]: You can always rerun the job or force push the git tag

Comment (regadas) [reply .github/workflows/release.yml:11]: right that's what I meant. I understand that this is probably a preference thing but i don't feel that fetching artifacts every time offer a lot of guarantees. If a build is green on `develop` for a given hash, running the release job (`develop` or `tag`) should mean that it will equally build and not be subjective to transient errors. Publishing to sonatype should be the only thing that can fail (which fails quite often unfortunately)

**Review by johnynek — APPROVED** [review]
This is great and super exciting!

One thing we probably want to address: the CI is a bit flaky on two kinds of tests: 1. approximate laws related to floating point numbers, 2. tests checking probabilistic bounds (this is true with Prob > y, which we run many trials and see if the result is highly improbable given that).

We should probably tune it so that those rarely happen.

**Comment by olafurpg** [comment]
Merging to see what happens on master. Small heads up, it usually takes a few failed attempts before it works correctly

**Comment by olafurpg** [comment]
The SNAPSHOT release worked, you can download the artifacts with
```
cs fetch com.twitter:algebird-core_2.11:0.13.7+37-fd6454ad-SNAPSHOT -r sonatype:snapshots
```
The next step is to try a stable release by pushing a git tag. Any objections if I push a v0.13.8-M1 tag to see if everything works?

FWIW, I have a `git-release` script in my `$PATH` 
```
❯ cat `which git-release`
set -eux
version=$1
git tag -af "v$version" -m "v$version" && git push -f origin v$version
```

I normally cut releases by running `git release 0.13.8` from the terminal. I usually wait until the CI release job has completed before publishing a release on GitHub.

**Comment by olafurpg** [comment]
> One thing we probably want to address: the CI is a bit flaky on two kinds of tests: 1. approximate laws related to floating point numbers, 2. tests checking probabilistic bounds (this is true with Prob > y, which we run many trials and see if the result is highly improbable given that).

The CI release currently runs the publish step without testing so flaky failures should not block the release. It's possible to add custom steps before `+publishSigned` in case you prefer that, but when I tried this in the past I normally ended up spending more time dealing with incomplete releases than catching actual bugs before a release (the master branch has presumably gone through PR code review).

**Comment by regadas** [comment]
>  v0.13.8-M1

we can go with a v0.13.8 since there are dependency updates, #844, #829 and some other minor things.

> It's possible to add  custom steps before +publishSigned

Yeah we can always just add `sbt +test` if needed but +1 on leaving it as is for now.

**Comment by regadas** [comment]
Let's include #850 as well

**Comment by regadas** [comment]
I think we are good for a new release `v0.13.8`. 

who should do the honours? I can prepare the release notes if you want /cc @johnynek @olafurpg

**Comment by johnynek** [comment]
One argument is the honors should go to @olafurpg who once again did the important work of making releases easier, and helping remove friction from the ecosystem. My second nomination would be @regadas who has really stepped up to do a lot of work managing these libraries. For the first time you will be able to release since you don't need the twitter sonatype credentials.

I'm happy to do it as well... :)

**Comment by olafurpg** [comment]
I pushed a tag for v0.13.8-M1 and it seems to have worked
```
❯ cs resolve com.twitter:algebird-core_2.12:0.13.8-M1
com.googlecode.javaewah:JavaEWAH:1.1.7:default
com.twitter:algebird-core_2.12:0.13.8-M1:default
org.scala-lang:scala-library:2.12.12:default
org.scala-lang:scala-reflect:2.12.12:default
org.scala-lang.modules:scala-collection-compat_2.12:2.2.0:default
org.typelevel:algebra_2.12:2.0.0:default
org.typelevel:cats-kernel_2.12:2.0.0:default
```

I opened #871 to make it easier to write release notes. I'll follow up with similar changes in the chill and bijection repos, although it may take a few days until I get around to it.

**Comment by regadas** [comment]
it worked! 🎉 however i thought you were releasing `v0.13.8`

**Comment by regadas** [comment]
@johnynek thx for the acknowledgement 🙏

**Comment by olafurpg** [comment]
You can do the non-milestone release, I just wanted to validate the pipeline was working :)

**Comment by olafurpg** [comment]
Followup PRs

* https://github.com/twitter/chill/pull/507
* https://github.com/twitter/bijection/pull/460

**Pull request merged**
